aarch64: pm_cpu_ops: Introduce pm_cpu_ops subsystem

carlocaione opened the pull request:
AArch64 has support for PSCI. This is especially useful for SMP because
PSCI is used to power on the secordary cores.

When the PSCI driver was introduced in Zephyr it was designed to rely on
a very PSCI-centric subsystem / interface.

There are two kinds of problems with this choice:

1. PSCI is only defined for the non-secure world and it is designed to
   boot CPU cores into non-secure state (that means that PSCI is only
   supposed to work if Zephyr is running in non-secure state)

2. There can be other ways or standards used to start / stop a core
   different from PSCI

This patch is trying to fix the original wrong assumption by making the
interface / subsystem a generic one, called 'pm_cpu_ops', and using PSCI
only as an actual driver that is a user of this new interface /
subsystem.

For now the new subsystem is only exposing two methods: cpu_on and
cpu_off, others will probably follow according to the needs.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

sandeepbrcm commented on drivers/pm_cpu_ops/Kconfig:
```diff
@@ -0,0 +1,19 @@
+module-str = pm_cpu_ops
+source "subsys/logging/Kconfig.template.log_config"
+
+source "drivers/pm_cpu_ops/Kconfig.psci"
```
something is still coupling with psci.  I think we should have global ops structure to register the ops. User / API need not worry about the underlying mechanism (psci or someother). For example RISC-V will register its SBI ops and someone may register local functions.

carlocaione replied:
What you think is coupling with PSCI? The `pm_cpu_ops` subsystem is now totally agnostic. The PSCI implementation is only using this agnostic interface. Here we are sourcing the PSCI Kconfig file.

sandeepbrcm commented on tests/arch/arm64/arm_psci/src/main.c:
```diff
@@ -24,12 +25,8 @@ void test_psci_func(void)
-
 	/* This should return -PSCI_RET_ALREADY_ON that is mapped to -EINVAL */
-	ret = psci_cpu_on(psci, 0, 0);
+	ret = pm_cpu_on(psci, 0, 0);
```
To operate on cpu what the user has to do ?
Get the device handle to a "pm compatible device"  which can be PSCI or something else.
and operate on it.
Instead should it just operate on the apis like `pm_cpu_ops.cpu_on(cpuid)`

The change required will be to maintain `pm_cpu_ops`  struct and provide registration api. 
The ops provider may or may not be a device. choosing the ops provider is a platform config choice. CONFIG_PSCI etc. The is conventional approach.

I want to see pros in this new approach as well if we have to deviate from the conventional approach. Thoughts?

carlocaione replied:
> To operate on cpu what the user has to do ?
> Get the device handle to a "pm compatible device" which can be PSCI or something else.
> and operate on it.

Correct.

> Instead should it just operate on the apis like `pm_cpu_ops.cpu_on(cpuid)`
> 
> The change required will be to maintain `pm_cpu_ops` struct and provide registration api.
> The ops provider may or may not be a device. choosing the ops provider is a platform config choice. CONFIG_PSCI etc. The is conventional approach.
> 
> I want to see pros in this new approach as well if we have to deviate from the conventional approach. Thoughts?

Another hybrid approach can be something like you proposed but leveraging the DTS `enable-method` of the cpu nodes property to register the proper driver / mechanism.

sandeepbrcm commented on CODEOWNERS:
```diff
@@ -244,7 +244,7 @@
 /drivers/pinmux/*hsdk*                    @iriszzw
 /drivers/pinmux/*it8xxx2*                 @ite
-/drivers/psci/                            @carlocaione
+/drivers/pm_cpu_ops/                      @carlocaione
```
now this is a generic cpu ops subsystem . architecture agnostic.  +1

carlocaione commented on drivers/pm_cpu_ops/CMakeLists.txt:
```diff
@@ -0,0 +1,6 @@
+zephyr_library()
+
+zephyr_library_sources_ifdef(CONFIG_ARM_PSCI	pm_cpu_ops_psci.c)
+zephyr_library_sources_ifdef(CONFIG_USERSPACE	pm_cpu_ops_handlers.c)
```
Not sure what you mean. The handlers are selectively compiled on `CONFIG_USERSPACE`. Where you referring to something different?

sandeepbrcm replied:
`CONFIG_USERSPACE `  is not relevant to cpu_ops right?

carlocaione replied:
well, it depends if you want to give userspace the possibility to shut down a core for example. I'm not strongly opinionated on this so I just decided to create the handlers.

carlocaione replied:
I'll take this out in the next respin.

carlocaione committed:
aarch64: pm_cpu_ops: Introduce pm_cpu_ops subsystem

AArch64 has support for PSCI. This is especially useful for SMP because
PSCI is used to power on the secordary cores.

When the PSCI driver was introduced in Zephyr it was designed to rely on
a very PSCI-centric subsystem / interface.

There are two kinds of problems with this choice:

1. PSCI is only defined for the non-secure world and it is designed to
   boot CPU cores into non-secure state (that means that PSCI is only
   supposed to work if Zephyr is running in non-secure state)

2. There can be other ways or standards used to start / stop a core
   different from PSCI

This patch is trying to fix the original wrong assumption by making the
interface / subsystem a generic one, called 'pm_cpu_ops', and using PSCI
only as an actual driver that is a user of this new interface /
subsystem.

For now the new subsystem is only exposing two methods: cpu_on and
cpu_off, others will probably follow according to the needs.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione commented:
@sandeepbrcm I think I have addressed your suggestions. The interface now is not tied to a device but it's a system-wide interface.

jharris-intel commented on include/drivers/pm_cpu_ops.h:
```diff
@@ -0,0 +1,63 @@
+/**
+ * @brief Power down the calling core
+ *
+ * This call is intended foruse in hotplug. A core that is powered down by
```
```suggestion
 * This call is intended for use in hotplug. A core that is powered down by
```

carlocaione replied:
ACK

jharris-intel commented on include/drivers/pm_cpu_ops.h:
```diff
@@ -0,0 +1,63 @@
+ * @retval 0 on success, a negative errno otherwise
+ * @retval -ENOTSUP If the operation is not supported
+ */
+int pm_cpu_on(unsigned long cpuid, unsigned long entry_point);
```
```
typedef void (entry_f)(void);
[...]
int pm_cpu_on(unsigned long cpuid, entry_f *entry_point);
```

...or at the very least `uintptr_t` please for an address.

carlocaione replied:
Agree. Fixed (using `uintptr_t`).

MrVan commented:
> AArch64 has support for PSCI. This is especially useful for SMP because
> PSCI is used to power on the secordary cores.
> 
> When the PSCI driver was introduced in Zephyr it was designed to rely on
> a very PSCI-centric subsystem / interface.
> 
> There are two kinds of problems with this choice:
> 
> 1. PSCI is only defined for the non-secure world and it is designed to
>    boot CPU cores into non-secure state (that means that PSCI is only
>    supposed to work if Zephyr is running in non-secure state)
> 2. There can be other ways or standards used to start / stop a core
>    different from PSCI
> 
> This patch is trying to fix the original wrong assumption by making the
> interface / subsystem a generic one, called 'pm_cpu_ops', and using PSCI
> only as an actual driver that is a user of this new interface /
> subsystem.
> 
> For now the new subsystem is only exposing two methods: cpu_on and
> cpu_off, others will probably follow according to the needs.
> 

So you are going to support spin table for ARM64? Should we force PSCI for ARM64?

carlocaione commented:
> So you are going to support spin table for ARM64?

I'm not personally interested in spin tables but in that case we can hook that up to this interface.

> Should we force PSCI for ARM64?

You do not always want PSCI on ARM64. For example you do not want that when you do not have a capable firmware running in EL2 or EL3 or when booting in secure mode. Or in general you can have other ways to bring your CPUs up (IIRC @jharris-intel is using some custom method instead of PSCI). So, no, the idea is not to force that.

MrVan commented on CODEOWNERS:
```diff
@@ -244,7 +244,7 @@
 /drivers/pinmux/*hsdk*                    @iriszzw
 /drivers/pinmux/*it8xxx2*                 @ite
-/drivers/psci/                            @carlocaione
+/drivers/pm_cpu_ops/                      @carlocaione
```
pm_cpu_ops seems not a good name to me. PSCI will be used for power management also, such as suspend/resume.
I would keep psci still as it is or like Linux Kernel use a firmware directory.

carlocaione replied:
This is exactly the opposite of what we are trying to do here. PSCI is just one way to perform CPU power management, but it's not the only way. To be able to support multiple implementations we have to create a "generic" interface for the CPU power management. PSCI is then using this generic interface to provide its services.

For example in `drivers/gpio` you have all the drivers using / implementing the GPIO interface, you do not have a directory for each implementation. The same here for `pm_cpu_ops`, in this directory you will find all the different users of this interface, including PSCI but possibly others as well.

About the name: `pm_cpu_ops` is literally taken from power management CPU operations. That can be a bit ugly but definitely better than `firmware` that is way too generic and broad.

About suspend/resume you are right but this is not something we currently support so I decided for now to leave those out. We can add the hooks to the interface when we will have something using them.

sandeepbrcm replied:
Re, the name `pm_cpu_ops`  you may use if anything more abstract pops up. So far I think this is appropriate. And the list of ops can be extended to cover all `known` cpu ops later from all architecture.
